src/bitops.c

```diff
@@ -242,7 +242,7 @@ long long serverPopcount(void *s, long count) {
 #if HAVE_X86_SIMD
     /* If length of s >= 256 bits and the CPU supports AVX2,
      * we prefer to use the SIMD version */
-    if (count >= 32) {
+    if (count >= 32 && __builtin_cpu_supports("avx2")) {
```

Comment:
since I do not expect avx2 support to be dynamically disabled on a CPU, you could use static storage for that and only calculate once when this function is called

```suggestion
    static bool cpu_supports_avx2 = __builtin_cpu_supports("avx2");
    if (count >= 32 && cpu_supports_avx2) {
```

Reply:
Doesn't work. I tested this with smaller file.

```
error: initializer element is not constant
    5 |     static bool cpu_supports_avx2 = __builtin_cpu_supports("avx2");
      |                                     ^~~~~~~~~~~~~~~~~~~~~~
```

Reply:
It seems the builtin results in `test  BYTE PTR  __cpu_model[rip+13], 4` (https://godbolt.org/z/oojsWfoEj), so the assembly expansion of anything `__cpuid` -related doesn't seem to be there. I can't say if this `test` is more expensive than using a [`call_once`](https://en.cppreference.com/w/c/thread/ONCE_FLAG_INIT.html) construct like this: https://godbolt.org/z/3fbzh9PM5.

Reply:
Interesting. Yes, I believe the CPU id structures are initialized once (by [`__builtin_cpu_init()`](https://gcc.gnu.org/onlinedocs/gcc/x86-Built-in-Functions.html#index-_005f_005fbuiltin_005fcpu_005finit-1)) to make the runtime checks faster.

Reply:
One last idea: We could split the function in multiple implementations to get rid of the runtime check and initialize a function pointer instead: https://godbolt.org/z/713srTrcb
I'm unsure if dereferencing that is cheaper than the `test` or `cmp` though.

Reply:
That's basically the IFUNC resolver approach. We did it for `string2ll()` in #2099 because it's called many times for every command, e.g. when parsing the RESP command arguments.

Since we need to backport this fix, it's good to keep it minimal.

We can perhaps do the resolver function in a separate PR, if the performance gain is worth the added complexity.

Reply:
yes. we can wait with that.

also we can just calculate it once in a constructor function and it would also be helpful for other cases:

```
static bool cpu_supports_avx2;

__attribute__((constructor))
static void init_cpu_flags(void) {
    cpu_supports_avx2 = __builtin_cpu_supports("avx2");
}
```

or call it from main...

Reply:
Oh @TedLyngmo I see you already referred to this proposal.

Reply:
`__builtin_cpu_supports("avx2")` expands to the same assembly as checking a single variable, so it's just as cheap. It can be seen in the godbolt links posted above.